cpp/src/dataframe/types.hpp

```diff
@@ -38,74 +39,161 @@
 /* ----------------------------------------------------------------------------*/
 namespace cudf
 {
-  struct category
+
+namespace detail
+{
+/**
+     * @brief Base wrapper structure to emulate "strong typedefs" for gdf_dtype values 
+     * that do not correspond to fundamental types.
+     * 
+     * Implements operators that allow the wrapper to be used as if it were a fundamental
+     * type.
+     * 
+     * @tparam T  The type of the wrapped value
+     * @tparam type_id  The wrapped gdf_dtype
+     */
+template <typename T, gdf_dtype type_id>
```

Comment:
1. Why do we need two template parameters? One of them is derivable from the other.
2. I'm not sure the wrapper class should "know" about `gdf_dtype`s at all; freestanding functions and out-of-class data structures or traits should be sufficient.

Reply:
This was primarily added per your request that the wrappers should have "code as documentation" that shows what corresponding `gdf_dtype` each of the wrappers wrap.

I'm not sure how either type is derivable from the other unless you use some sort of traits that map a `gdf_dtype` to a `T`, but I don't see why that's better than just having 2 template parameters.

Reply:
The wrapper class in itself, as implemented, has nothing in particular to do with cuDF columns and their types. It's completely generic.  

You can still use a "code as documentation" approach to relate specific wrappers to `gdf_dtype`, in one of two ways I can think of:

1. Use the traits class to translates `gdf_dtype` values to wrapper types, and the traits  That documents exactly what we want. And if any one of them already exists - then the documentation is already there.
2. Keep your structs instead of `using` statements, and add a `corresponding_column_type` enum, or static constexpr, to each of those structs, with the appropriate values.